app/models/miq_task.rb

```diff
@@ -33,12 +33,21 @@ class MiqTask < ApplicationRecord
     t.grouping(Arel::Nodes::Case.new(t[:state]).when(STATE_FINISHED).then(t[:status]).else(t[:state]))
   end)
 
-  scope :active,                  ->          { where(:state => STATE_ACTIVE) }
-  scope :no_associated_job,       ->          { where.not("id IN (SELECT miq_task_id from jobs)") }
-  scope :timed_out,               ->          { where("updated_on < ?", Time.now.utc - ::Settings.task.active_task_timeout.to_i_with_method) }
-  scope :with_userid,             ->(userid)  { where(:userid => userid) }
-  scope :with_zone,               ->(zone)    { where(:zone => zone) }
-  scope :with_updated_on_between, ->(range)   { where(:updated_on => range) }
+  scope :active,                  ->           { where(:state => STATE_ACTIVE) }
+  scope :no_associated_job,       ->           { where.not("id IN (SELECT miq_task_id from jobs)") }
+  scope :timed_out,               ->           { where("updated_on < ?", Time.now.utc - ::Settings.task.active_task_timeout.to_i_with_method) }
+  scope :with_userid,             ->(userid)   { where(:userid => userid) }
+  scope :with_zone,               ->(zone)     { where(:zone => zone) }
+  scope :with_updated_on_between, ->(from, to) { where("miq_tasks.updated_on BETWEEN ? AND ?", from, to) }
+  scope :with_state,              ->(state)    { where(:state => state) }
+  scope :finished,                ->           { with_state('Finished') }
+  scope :running,                 ->           { where.not(:state => %w(Finished Waiting_to_start Queued)) }
+  scope :queued,                  ->           { with_state(%w(Waiting_to_start Queued)) }
+  scope :completed_ok,            ->           { finished.where(:status => 'Ok') }
+  scope :completed_warn,          ->           { finished.where(:status => 'Warn') }
+  scope :completed_error,         ->           { finished.where(:status => 'Error') }
+  scope :no_status_selected,      ->           { running.where.not(:status => %(Ok Error Warn)) }
+  scope :with_status_in,          ->(scopes)   { scopes[1..-1].reduce(MiqTask.send(scopes.first)) { |chain, s| chain.or(MiqTask.send(s)) } }
```

Comment:
This is not the most handsome scope out there, I have to admit that. It allows us to chain scopes to get a query like:
```
scope1 AND scope2 AND (scope3 OR scope4 OR scope5)
```
using scopes (matching to example above)
```
:named_scopes => [:scope1, :scope2, [:with_status_in, [:scope3, :scope4, :scope5]]]
```

Reply:
This is an alternative to what previously was an SQL fragment passed around. So while this is not perfect, it's an improvement on the current situation.

Reply:
The later patch uses splat instead of slices, which is a bit better code...